### ISSUE TITLE
systemd's PermissionsStartOnly is deprecated

### ISSUE THREAD

**Issue by nh2** [NixOS/nixpkgs#53852]
See https://github.com/systemd/systemd/pull/10802:

`PermissionsStartOnly` is deprecated ([but not yet removed](https://github.com/systemd/systemd/pull/10802#issuecomment-439446299)); the replacement are the `Special executable prefixes` mentioned in https://www.freedesktop.org/software/systemd/man/systemd.service.html#ExecStart=.

We have many occurrences this in nixpkgs (as per commit ad23ea312d37b8086ea62bd6be33c2c07e5a9963):

```
% git grep PermissionsStartOnly upstream/master | wc -l
130
```

It's not urgent, as it still works right now, but it's problematic that you can't any longer find any docs about a feature that we use almost everywhere.

---

Fixing it isn't totally trivial because `PermissionsStartOnly` is/was a separate setting, but with the new approach we actually have to write characters in front of `ExecStartPre` lines and so on.

**Edit:** Another key problem is that `serviceConfig` is an attrset, so you cannot use it to do what systemd allows: Creating multiple `ExecStart` lines (or similar lines) like:

```
ExecStart=+...something here...
ExecStart=...something else here...
```

because in an attrset there can be only 1 `ExecStart`. **Edit:** Likely workaround for that in https://github.com/NixOS/nixpkgs/issues/53852#issuecomment-751409369.

**Comment by nh2** [comment]
For context, here are the docs that are now no longer there:

> `PermissionsStartOnly`
>
> Takes a boolean argument. If true, the permission-related execution options, as configured with User= and similar options (see `systemd.exec(5)` for more information), are only applied to the process started with `ExecStart=`, and not to the various other `ExecStartPre=`, `ExecStartPost=`, `ExecReload=`, `ExecStop=`, and `ExecStopPost=` commands. If `false`, the setting is applied to all configured commands the same way. Defaults to `false`.

**Comment by flokli** [comment]
@nh2 in a lot of cases, the `ExecPreStart` scripts only set up a directory structure, which should probably better be done by `systemd.tmpfiles.rules`

**Comment by fpletz** [comment]
The change looks easy enough. I propose we target 19.09 to deprecate this option in the NixOS systemd module with a warning and fix all of our modules until then.

**Comment by nh2** [comment]
@fpletz That sounds good to me.

**Comment by aanderse** [comment]
@nh2 @flokli @fpletz  I don't mind picking away at this, but I'd like to make sure I'm on the right track... does https://github.com/NixOS/nixpkgs/pull/56265 look good so far? I figured commit per module so it can easily be cherry picked.

**Comment by nh2** [comment]
@aanderse From a quick look, the approach looks correct to me.

I'd mention this issue in each commit message though so that people can easily find this issue from them.

**Comment by bricewge** [comment]
How to fix services like the one fot hydron? https://github.com/NixOS/nixpkgs/blob/9fa52ae9a22091c978857ef5440372c06da4c99e/nixos/modules/services/web-servers/hydron.nix#L95-L123
It seems that to remove `PermissionsStartOnly = true` we need to replace `ExecStartPre=/nix/store/...-unit-script-hydron-pre-start` by `ExecStartPre=+/nix/store/...-unit-script-hydron-pre-start` (notice the added *+*), how can this be achived?
I'm asking because I have this issue in writing a module for miniflux.

**Comment by aanderse** [comment]
@bricewge I *would* suggest https://github.com/NixOS/nixpkgs/pull/56720, but that PR isn't ready quite yet... I guess I should get that finished and merged.

**Comment by joachifm** [comment]
@zimbatm I agree that tmpfiles is an improper replacement for `StateDirectory`, at least conceptually ... `StateDirectory` can be set per unit, but I take it the motivation for using tmpfiles is greater control over mode and ownership bits. From reading the manual, however, it seems to me that setting `User` and `Group` suffices to create dirs & ensure they are owned by the user/group, which I'd imagine suffices for a lot of (most?) cases, but I could be missing something ...

**Comment by flokli** [comment]
From 

> Dropping the link to an old comment here here, in case it gets lost again during rebases:
>
> https://github.com/NixOS/nixpkgs/pull/56265#discussion_r264064056
>
> TLDR:
> tmpfiles is a very unlucky name - these directories are really not temporary, and could be used if `StateDirectory`, `RuntimeDirectory` and `ConfigurationDirectory` aren't enough, (or more complex directory layouts inside need to be created or shared across multiple units)

We should use `StateDirectory`, `RuntimeDirectory` and `ConfigurationDirectory` where possible, but sometimes it's not enough (when files are shared between multiple units, symlinks needed to be set up, custom permissions etc.

**Comment by xaverdh** [comment]
`rg --files-with-matches PermissionsStartOnly | wc -l` outputs 37 on master, so not stale

**Comment by aanderse** [comment]
Help wanted, but I do *intend* to continue chipping away at this eventually.

**Comment by nh2** [comment]
I'm trying to package a Django app, and need to make its `/static/` subdir (which is in `/var`, not in the nix store) world-readable recursively so that nginx can read it.

How do I do that with `systemd-tmpfiles`?

**Comment by nh2** [comment]
> **Edit:** Another key problem is that `serviceConfig` is an attrset, so you cannot use it to do what systemd allows: Creating multiple `ExecStart` lines


> I'm trying to package a Django app, and need to make its `/static/` subdir (which is in `/var`, not in the nix store) world-readable recursively so that nginx can read it.



I think a workaround for that is to use 1 `ExecStart` line and separate the invocations by ` ; ` (with spaces left and right of it). That way, `+` can be used to run some of those invocations as root. [Example from here](
https://github.com/nh2/u51/blob/4cfa62aef1f08b3271cc701072c289499aa52863/nix/nixos/modules/services/u51.nix#L78):

```
ExecStart = "+${u51PermsScript} ; ${managepyScript}";
```

**Comment by flokli** [comment]
> I'm trying to package a Django app, and need to make its `/static/` subdir (which is in `/var`, not in the nix store) world-readable recursively so that nginx can read it.
> 
> How do I do that with `systemd-tmpfiles`?

If the user specificed in the unit owns the files in `/static`, it should be able to just `chmod -R +x`. shouldn't it? You don't need to become root for this.

**Comment by flokli** [comment]
This is not stale. There's still a lot of modules using `PermissionsStartOnly`, which need to be migrated.

**Comment by RaitoBezarius** [comment]
For NixOS 23.05:

```
❯ git grep PermissionsStartOnly | wc -l
37
```

Based on the maintenance data in there:

- go-neb: @mweinelt @maralorn 
- matomo: @florianjacob 
- neo4j: @patternspandemic @jonringer @erictapen 
- foundationdb: @thoughtpolice 
- tomcat: @danbst 
- dkimproxy-out: @Ekleog 
- blackfire: @shyim 

This would bring us to 30 with non-explicit maintainers on them, I will try to look in the Git history when I have time.

**Comment by shyim** [comment]
blackfire explictly removes it `PermissionsStartOnly` https://github.com/NixOS/nixpkgs/blob/5af1d31e42c8e0bdf6c6d9c2879fab832dea4a61/pkgs/development/tools/misc/blackfire/default.nix#L39

I guess that should be fine :)

**Comment by RaitoBezarius** [comment]
> blackfire explictly removes it `PermissionsStartOnly`
> 
> https://github.com/NixOS/nixpkgs/blob/5af1d31e42c8e0bdf6c6d9c2879fab832dea4a61/pkgs/development/tools/misc/blackfire/default.nix#L39
> 
> I guess that should be fine :)

Shame on me! :) Thank you!

**Comment by nh2** [comment]
Relevant: https://www.phoronix.com/news/systemd-tmpfiles-purge-drama

The fact that something called `tmpfiles` is used to store very important permanent non-temp files is quite confusing.

The syntax is also not great:

```nix
        "d '${cfg.settings.dbDir}' 0700 ${cfg.user} ${cfg.group} - -"
```

Single-letter characters, random minuses, all arguments everything by position ... this is not a good design for readability, understanding, or correctness.
